File: src/main/java/com/github/sorusclient/client/module/impl/togglesprintsneak/ToggleSprintSneak.java
package com.github.sorusclient.client.module.impl.togglesprintsneak;

import com.github.sorusclient.client.Sorus;
import com.github.sorusclient.client.adapter.IAdapter;
import com.github.sorusclient.client.adapter.IKeyBind;
import com.github.sorusclient.client.adapter.Key;
import com.github.sorusclient.client.adapter.event.KeyEvent;
import com.github.sorusclient.client.event.EventManager;
import com.github.sorusclient.client.module.ModuleDisableable;
import com.github.sorusclient.client.setting.Setting;
import com.github.sorusclient.client.setting.ConfigurableData;

import java.util.List;

public class ToggleSprintSneak extends ModuleDisableable {

    private final Setting<Boolean> toggleSprint;
    private final Setting<Boolean> useCustomSprintKey;
    private final Setting<Key> customSprintKey;
    private final Setting<Boolean> toggleSneak;
    private final Setting<Boolean> useCustomSneakKey;
    private final Setting<Key> customSneakKey;

    private boolean sprintToggled;
    private boolean sneakToggled;

    public ToggleSprintSneak() {
        super("toggleSprintSneak");

        this.register("toggleSprint", this.toggleSprint = new Setting<>(false));
        this.register("useCustomSprintKey", this.useCustomSprintKey = new Setting<>(false));
        this.register("customSprintKey", this.customSprintKey = new Setting<>(Key.SHIFT_LEFT));
        this.register("toggleSneak", this.toggleSneak = new Setting<>(false));
        this.register("useCustomSneakKey", this.useCustomSneakKey = new Setting<>(false));
        this.register("customSneakKey", this.customSneakKey = new Setting<>(Key.CONTROL_LEFT));

        Sorus.getInstance().get(EventManager.class).register(KeyEvent.class, this::onKey);
    }

    private void onKey(KeyEvent event) {
        if (event.isPressed() && !event.isRepeat()) {
            if (event.getKey() == this.getSprintKey()) {
                this.sprintToggled = !this.sprintToggled;
            }

            if (event.getKey() == this.getSneakKey()) {
                this.sneakToggled = !this.sneakToggled;
            }
        }
    }

    private Key getSprintKey() {
        IAdapter adapter = Sorus.getInstance().get(IAdapter.class);
        return this.useCustomSprintKey.getValue() ? this.customSprintKey.getValue() : adapter.getKeyBind(IKeyBind.KeyBindType.SPRINT).getKey();
    }

    private Key getSneakKey() {
        IAdapter adapter = Sorus.getInstance().get(IAdapter.class);
        return this.useCustomSneakKey.getValue() ? this.customSneakKey.getValue() : adapter.getKeyBind(IKeyBind.KeyBindType.SNEAK).getKey();
    }

    public boolean isSprintToggled() {
        return this.isEnabled() && this.toggleSprint.getValue() && this.sprintToggled;
    }

    public boolean isSneakToggled() {
        return this.isEnabled() && this.toggleSneak.getValue() && this.sneakToggled;
    }

    @Override
    public void addSettings(List<ConfigurableData> settings) {
        super.addSettings(settings);
        settings.add(new ConfigurableData.Toggle("Toggle Sprint", this.toggleSprint));
        settings.add(new ConfigurableData.Toggle("Use Custom Sprint Key", this.useCustomSprintKey));
        settings.add(new ConfigurableData.KeyBind("Custom Sprint Key", this.customSprintKey));
        settings.add(new ConfigurableData.Toggle("Toggle Sneak", this.toggleSneak));
        settings.add(new ConfigurableData.Toggle("Use Custom Sneak Key", this.useCustomSneakKey));
        settings.add(new ConfigurableData.KeyBind("Custom Sneak Key", this.customSneakKey));
    }

}

File: src/main/kotlin/com/github/sorusclient/client/module/impl/togglesprintsneak/ToggleSprintSneak.kt
package com.github.sorusclient.client.module.impl.togglesprintsneak

import com.github.sorusclient.client.adapter.AdapterManager
import com.github.sorusclient.client.adapter.IKeyBind
import com.github.sorusclient.client.adapter.Key
import com.github.sorusclient.client.adapter.event.KeyEvent
import com.github.sorusclient.client.event.EventManager
import com.github.sorusclient.client.module.ModuleDisableable
import com.github.sorusclient.client.setting.ConfigurableData
import com.github.sorusclient.client.setting.ConfigurableData.KeyBind
import com.github.sorusclient.client.setting.ConfigurableData.Toggle
import com.github.sorusclient.client.setting.Setting

class ToggleSprintSneak : ModuleDisableable("toggleSprintSneak") {
    private var toggleSprint: Setting<Boolean>
    private var useCustomSprintKey: Setting<Boolean>
    private var customSprintKey: Setting<Key>
    private var toggleSneak: Setting<Boolean>
    private var useCustomSneakKey: Setting<Boolean>
    private var customSneakKey: Setting<Key>
    private var sprintToggled = false
    private var sneakToggled = false

    init {
        register("toggleSprint", Setting(false).also { toggleSprint = it })
        register("useCustomSprintKey", Setting(false).also { useCustomSprintKey = it })
        register("customSprintKey", Setting(Key.SHIFT_LEFT).also { customSprintKey = it })
        register("toggleSneak", Setting(false).also { toggleSneak = it })
        register("useCustomSneakKey", Setting(false).also { useCustomSneakKey = it })
        register("customSneakKey", Setting(Key.CONTROL_LEFT).also { customSneakKey = it })
        EventManager.register(this::onKey)
    }

    private fun onKey(event: KeyEvent) {
        if (event.isPressed && !event.isRepeat) {
            if (event.key == sprintKey) {
                sprintToggled = !sprintToggled
            }
            if (event.key == sneakKey) {
                sneakToggled = !sneakToggled
            }
        }
    }

    private val sprintKey: Key
        get() {
            val adapter = AdapterManager.getAdapter()
            return if (useCustomSprintKey.value) customSprintKey.value else adapter.getKeyBind(IKeyBind.KeyBindType.SPRINT).key
        }
    private val sneakKey: Key
        get() {
            val adapter = AdapterManager.getAdapter()
            return if (useCustomSneakKey.value) customSneakKey.value else adapter.getKeyBind(IKeyBind.KeyBindType.SNEAK).key
        }

    fun isSprintToggledValue(): Boolean {
        return isEnabled() && toggleSprint.value && sprintToggled
    }

    fun isSneakToggledValue(): Boolean {
        return isEnabled() && toggleSneak.value && sneakToggled
    }

    override fun addSettings(settings: MutableList<ConfigurableData>) {
        super.addSettings(settings)
        settings.add(Toggle("Toggle Sprint", toggleSprint))
        settings.add(
            ConfigurableData.Dependent(
                Toggle("Use Custom Sprint Key", useCustomSprintKey),
                toggleSprint,
                true
            )
        )
        settings.add(
            ConfigurableData.Dependent(
                ConfigurableData.Dependent(
                    KeyBind(
                        "Custom Sprint Key",
                        customSprintKey
                    ), useCustomSprintKey, true
                ), toggleSprint, true
            )
        )
        settings.add(Toggle("Toggle Sneak", toggleSneak))
        settings.add(ConfigurableData.Dependent(Toggle("Use Custom Sneak Key", useCustomSneakKey), toggleSneak, true))
        settings.add(
            ConfigurableData.Dependent(
                ConfigurableData.Dependent(
                    KeyBind("Custom Sneak Key", customSneakKey),
                    useCustomSneakKey,
                    true
                ), toggleSneak, true
            )
        )
    }
}